packages/material-ui-lab/src/MasonryItem/MasonryItem.js

```diff
@@ -0,0 +1,174 @@
+      // all contents should have a width of 100%
+      width: '100%',
+      boxSizing: 'inherit',
+    },
+    visibility: styleProps.height ? 'visible' : 'hidden',
+    gridColumnEnd: `span ${styleProps.columnSpan}`,
+    boxSizing: 'inherit',
+  };
+
+  const base = {};
+  Object.keys(theme.breakpoints.values).forEach((breakpoint) => {
+    if (styleProps.spacing[breakpoint] != null) {
+      base[breakpoint] = true;
+    }
+  });
+  const spacingValues = resolveBreakpointValues({ values: styleProps.spacing, base });
+  const transformer = createUnarySpacing(theme);
+  const styleFromPropValue = (propValue) => {
+    const gap = styleProps.height ? Number(getValue(transformer, propValue).replace('px', '')) : 0;
+    // for lazy-loaded images to load properly, they should take up space greater than 1px
+    // taking into account a row gap of 1px, rowSpan should at least be 3
+    const rowSpan = styleProps.height ? Math.ceil(styleProps.height + gap) : 3;
+    return {
+      gridRowEnd: `span ${rowSpan}`,
+      paddingBottom: gap - 1,
+    };
+  };
+  styles = { ...styles, ...handleBreakpoints({ theme }, spacingValues, styleFromPropValue) };
+
+  return styles;
+};
+
+const MasonryItemRoot = styled('div', {
+  name: 'MuiMasonryItem',
+  slot: 'Root',
+  overridesResolver: (props, styles) => {
+    return [styles.root];
+  },
+})(style);
+
+const MasonryItem = React.forwardRef(function MasonryItem(inProps, ref) {
+  const props = useThemeProps({
+    props: inProps,
+    name: 'MuiMasonryItem',
+  });
+
+  const masonryItemRef = React.useRef(null);
+
+  const { spacing = 1 } = React.useContext(MasonryContext);
+  const { children, className, component = 'div', columnSpan = 1, height, ...other } = props;
+  const [styleProps, setStyleProps] = React.useState({
+    ...props,
+    spacing,
+    columnSpan,
+    height: height < 0 ? 0 : height, // MasonryItems to which negative or zero height is passed will be hidden
+  });
+
+  const classes = useUtilityClasses(styleProps);
+
+  React.useEffect(() => {
+    // If height is passed by user, ResizeObserver is not used.
+    if (height !== undefined) {
+      setStyleProps({
+        ...styleProps,
+        spacing,
+        columnSpan,
+        height: height < 0 ? 0 : height,
+      });
+
+      return true;
+    }
+    const resizeObserver = new ResizeObserver(([item]) => {
+      setStyleProps({
+        ...styleProps,
+        spacing,
+        columnSpan,
+        height: item.contentRect.height,
+      });
+    });
+    const handleError = () => {
+      const src = masonryItemRef.current.firstChild.src;
+      masonryItemRef.current.firstChild.src = src;
+    };
+    const item = masonryItemRef.current?.firstChild;
+    item.addEventListener('error', handleError);
+    resizeObserver.observe(item);
+    return () => {
+      resizeObserver.unobserve(item);
+      item.removeEventListener('error', handleError);
+    };
+  }, [height, spacing, columnSpan]); // eslint-disable-line
```

Comment:
This is still problematic. We need some more context here what you specifically disable and why you disable it.

Reply:
`styleProps` should be included in the dependency array for eslint to stop complaining, and this causes infinite re-rendering. I will put this in a comment next to the code.

Reply:
@hbjORbj I think that the point of Sebastian was to restructure the logic. the infinite loop is suspicious:

<img width="772" alt="Capture d’écran 2021-08-05 à 20 13 02" src="https://user-images.githubusercontent.com/3165635/128400325-752a55e6-0b00-4de9-b974-5c0b6d6fde05.png">

Reply:
I see. Thank you. It's fixed here: https://github.com/mui-org/material-ui/pull/27439/commits/b1c4cbc9521d587a65093c916144f30bb323fd09